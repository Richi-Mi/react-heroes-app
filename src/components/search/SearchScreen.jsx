import React, { Fragment, useMemo } from 'react';
import queryString from 'query-string';

// modulo query-string

import useForm from '../../hooks/useForm';
import '../../design/search.css';
import getHeroesByName from '../../selectors/getHeroesByName';
import HeroCard from '../hero/HeroCard';
import { useNavigate, useLocation } from 'react-router-dom';

const SearchScreen = () => {
    const navigate = useNavigate();
    const location = useLocation();

    const { search = '' } = queryString.parse( location.search );
    

    const [ formValues, handleInputSearch ] = useForm({
        searchText: search
    });

    const { searchText } = formValues
    let heroesFiltered = useMemo( () => {
        return getHeroesByName( search )
    }, [ search ]);
    
    const handleSubmit = ( ev ) => {
        ev.preventDefault();
        navigate(`?search=${ searchText }`)
    }
    
    return (
        <Fragment>
            <h1> Busquedas </h1>
            <hr />

            <form className="form_search" action='submit' onSubmit={ handleSubmit }>
                <div className="form_section">
                    <input
                        type="text"
                        placeholder='Buscar un Heroe'
                        className='form_input'
                        name='searchText'
                        autoComplete='off'
                        value={ searchText }
                        onChange={ handleInputSearch }
                    />
                </div>
                <div className="form_section_btn">
                    <button type='submit' className='btn_search'>
                        Search
                        <i className="fas fa-search"></i>
                    </button>
                </div>
            </form>

            <hr />

            <div className="results">
                <div>
                    <h3 className='title_results'> Resultados: </h3>
                    {
                        ( search === '' ) 
                            ? <p className='alert_info'> Busca un Heroe </p>
                            : ( heroesFiltered.length === 0 ) && <p className='alert_danger'> No hay resultados para: { search } </p>
                    }
                </div>
                <div className="cards_results">
                    {
                        heroesFiltered.map( hero => {
                            return <HeroCard key={ hero.id } {...hero} />
                        })
                    }
                </div>
            </div>

        </Fragment>
    )
};

export default SearchScreen;
