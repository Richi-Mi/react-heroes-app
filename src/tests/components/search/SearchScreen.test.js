import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import SearchScreen from "../../../components/search/SearchScreen";

const mockNavigate = jest.fn(); // Obligatorio poner mock para hacer funciones

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate 
}));

describe(' * Pruebas en el componente <SearchScreen /> ', () => {
    test(' - Debe de Mostrarse correctamente con valores por defecto', () => {
    
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search'] }>
                <SearchScreen />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert_info').text().trim() ).toBe('Busca un Heroe')
    });

    test(' - Debe de mostrar a Batman y el valor del QueryString', () => {
        const queryParam = 'batman'

        const wrapper = mount(
            <MemoryRouter initialEntries={ [`/search?search=${ queryParam }`] }>
                <SearchScreen />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('input').prop('value') ).toBe( queryParam );
    })

    test(' - Debe de mostrar un error si no encuentra un Heroe', () => {
        const queryParam = 'batman12343'

        const wrapper = mount(
            <MemoryRouter initialEntries={ [`/search?search=${ queryParam }`] }>
                <SearchScreen />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert_danger').text().trim() ).toBe(`No hay resultados para: ${ queryParam }`);
        expect( wrapper.find('.alert_danger').text().trim().includes( queryParam) ).toBe( true );      
    });
    
    test(' - Debe de llamar el navigate a la nueva pantalla', () => {

        const query = 'batman'
        const wrapper = mount(
            <MemoryRouter initialEntries={ [`/search`] }>
                <SearchScreen />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', { 
            target: { 
                value: query,
                name: 'searchText' 
            } 
        } );

        wrapper.find('form').simulate('submit', {
            preventDefault: () => {  }
        });

        expect( mockNavigate ).toHaveBeenCalled();
        expect( mockNavigate ).toHaveBeenCalledWith(`?search=${ query }`);
    });
    
    
});
