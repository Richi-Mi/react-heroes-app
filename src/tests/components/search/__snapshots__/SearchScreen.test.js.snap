// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` * Pruebas en el componente <SearchScreen />   - Debe de Mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h1>
     Busquedas 
  </h1>,
  <hr />,
  <form
    action="submit"
    className="form_search"
    onSubmit={[Function]}
  >
    <div
      className="form_section"
    >
      <input
        autoComplete="off"
        className="form_input"
        name="searchText"
        onChange={[Function]}
        placeholder="Buscar un Heroe"
        type="text"
        value=""
      />
    </div>
    <div
      className="form_section_btn"
    >
      <button
        className="btn_search"
        type="submit"
      >
        Search
        <i
          className="fas fa-search"
        />
      </button>
    </div>
  </form>,
  <hr />,
  <div
    className="results"
  >
    <div>
      <h3
        className="title_results"
      >
         Resultados: 
      </h3>
      <p
        className="alert_info"
      >
         Busca un Heroe 
      </p>
    </div>
    <div
      className="cards_results"
    />
  </div>,
]
`;

exports[` * Pruebas en el componente <SearchScreen />   - Debe de mostrar a Batman y el valor del QueryString 1`] = `
Array [
  <h1>
     Busquedas 
  </h1>,
  <hr />,
  <form
    action="submit"
    className="form_search"
    onSubmit={[Function]}
  >
    <div
      className="form_section"
    >
      <input
        autoComplete="off"
        className="form_input"
        name="searchText"
        onChange={[Function]}
        placeholder="Buscar un Heroe"
        type="text"
        value="batman"
      />
    </div>
    <div
      className="form_section_btn"
    >
      <button
        className="btn_search"
        type="submit"
      >
        Search
        <i
          className="fas fa-search"
        />
      </button>
    </div>
  </form>,
  <hr />,
  <div
    className="results"
  >
    <div>
      <h3
        className="title_results"
      >
         Resultados: 
      </h3>
    </div>
    <div
      className="cards_results"
    >
      <div
        className="card animate__animated animate__fadeIn"
      >
        <div
          className="card_img"
        >
          <img
            alt="Batman"
            src="/assets/dc-batman.jpg"
          />
        </div>
        <h3
          className="dc"
        >
           
          Batman
           
        </h3>
        <div
          className="card_info"
        >
          <div
            className="card_text"
          >
            <p>
               
              Bruce Wayne
               
            </p>
          </div>
          <div
            className="card_link"
          >
            <a
              className="link"
              href="/hero/dc-batman"
              onClick={[Function]}
            >
              Mas 
              <i
                className="fas fa-arrow-right"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[` * Pruebas en el componente <SearchScreen />   - Debe de mostrar un error si no encuentra un Heroe 1`] = `
Array [
  <h1>
     Busquedas 
  </h1>,
  <hr />,
  <form
    action="submit"
    className="form_search"
    onSubmit={[Function]}
  >
    <div
      className="form_section"
    >
      <input
        autoComplete="off"
        className="form_input"
        name="searchText"
        onChange={[Function]}
        placeholder="Buscar un Heroe"
        type="text"
        value="batman12343"
      />
    </div>
    <div
      className="form_section_btn"
    >
      <button
        className="btn_search"
        type="submit"
      >
        Search
        <i
          className="fas fa-search"
        />
      </button>
    </div>
  </form>,
  <hr />,
  <div
    className="results"
  >
    <div>
      <h3
        className="title_results"
      >
         Resultados: 
      </h3>
      <p
        className="alert_danger"
      >
         No hay resultados para: 
        batman12343
         
      </p>
    </div>
    <div
      className="cards_results"
    />
  </div>,
]
`;
